starfish/test/test_multiprocessing.py

```diff
@@ -0,0 +1,113 @@
+from multiprocessing import Array as mp_array  # type: ignore
+from typing import Any, Callable
+
+import numpy as np
+
+from starfish.multiprocessing import shmem
+
+
+def test_numpy_array(nitems: int=10):
+    """
+    Try to share a numpy array directly.  This should fail as numpy is by default a copy-on-write
+    object.  The worker process will write to array and we should not see the changes.
+    """
+    array = np.zeros(shape=(nitems,), dtype=np.uint8)
+    _start_process_to_test_shmem(array, _decode_numpy_array_to_numpy_array, nitems)
+    for ix in range(nitems):
+        assert array[ix] == 0
+
+
+def _decode_numpy_array_to_numpy_array(array):
+    return array
+
+
+def test_shmem_numpy_array(nitems: int=10):
+    """
+    Try to share a numpy array based on a multiprocessing Array object.  The array object is passed
+    to the worker process.  The worker process reconstitutes the numpy array from that memory buffer
+    and writes to the reconstituted numpy array.  Writes in the worker process should be visible in
+    the parent process.
+    """
+    buffer = mp_array(ctypes.c_uint8, nitems)
+    array = _decode_array_to_numpy_array(buffer)
+    array.fill(0)
+    _start_process_to_test_shmem(buffer, _decode_array_to_numpy_array, nitems)
+    for ix in range(nitems):
+        assert array[ix] == ix
+
+
+def _decode_array_to_numpy_array(array):
+    return np.frombuffer(array.get_obj(), dtype=np.uint8)
+
+
+class TestWrappedArray:
+    """
+    Dummy class that wraps a multiprocessing array in an object.
+    """
+    def __init__(self, array: mp_array) -> None:
+        self.array = array
+
+
+def test_wrapped_shmem_numpy_array(nitems: int=10):
+    """
+    Try to share a numpy array based on a multiprocessing Array object.  The array object is wrapped
+    in a container class and passed to the worker process.  The worker process reconstitutes the
+    numpy array from that memory buffer and writes to the reconstituted numpy array.  Writes in the
+    worker process should be visible in the parent process.
+    """
+    buffer = mp_array(ctypes.c_uint8, nitems)
+    array = _decode_array_to_numpy_array(buffer)
+    array.fill(0)
+    _start_process_to_test_shmem(
+        TestWrappedArray(buffer), _decode_wrapped_array_to_numpy_array, nitems)
+    for ix in range(nitems):
+        assert array[ix] == ix
+
+
+def _decode_wrapped_array_to_numpy_array(wrapped_array):
+    return np.frombuffer(wrapped_array.array.get_obj(), dtype=np.uint8)
+
+
+def _applied_func(decoder: Callable[[Any], np.ndarray], position: int) -> None:
+    array = decoder(shmem.get_payload())
+    array[position] = position
+
+
+def _start_process_to_test_shmem(
+        array_holder: Any,
+        decoder: Callable[[Any], np.ndarray],
+        nitems: int) -> None:
+    """
+    Starts a process and passes an object to its initializer.  The object is passed to
```

Comment:
Would replace docs with the following, which is a bit more verbose about the rationale for implementing this function. 

```
    This method leverages multiprocessing.Pool to test process initializers and 
    shared memory array decoders. 

    Starts a process Pool and passes :py:method:`shmem.initializer` as its initializer, 
    which will store `array_holder` as a global variable inside the initialized process. 
    
    Creates `bound_func`, which will be executed by the subprocesses of Pool. 
    `bound_func` is a partial function constructed from 
    :py:method:`_write_to_shared_memory_array` that exposes `decoder`, a method 
    that gets a numpy array out of array_holder and then writes a value to it 
    (inside the spawned process). 

    This enables writing of tests that check whether the written values are available
    in the parent processes, which depends on `shmem.initializer`, `array_holder`, 
    and `decoder`.  
```

Reply:
Applied.